tests/test_rlearner.py

```diff
@@ -26,3 +37,98 @@ def test_r_loss(use_pandas):
         propensity_scores=propensity_scores,
     )
     assert result == pytest.approx(2, abs=1e-4, rel=1e-4)
+
+
+@pytest.mark.parametrize(
+    "treatment_model_factory, onnx_converter",
+    (
+        list(
+            zip(
+                all_sklearn_regressors,
+                [convert_sklearn] * len(all_sklearn_regressors),
+            )
+        )
+        + [
+            (LGBMRegressor, convert_lightgbm),
+            (XGBRegressor, convert_xgboost),
+        ]
+    ),
+)
+@pytest.mark.parametrize("is_classification", [True, False])
+def test_rlearner_onnx(
+    treatment_model_factory, onnx_converter, is_classification, onnx_dataset
+):
+    if not function_has_argument(treatment_model_factory.fit, "sample_weight"):
+        pytest.skip()
+
+    supports_categoricals = treatment_model_factory in [
+        LGBMRegressor,
+        # convert_sklearn does not support categoricals https://github.com/onnx/sklearn-onnx/issues/1051
+        # HistGradientBoostingRegressor,
+        # convert_xgboost does not support categoricals https://github.com/onnx/onnxmltools/issues/469#issuecomment-1993880910
+        # XGBRegressor,
+    ]
+
+    X_numerical, X_with_categorical, y_class, y_reg, w = onnx_dataset
+    n_numerical_features = X_numerical.shape[1]
+
+    if supports_categoricals:
+        X = X_with_categorical
+        n_categorical_features = X.shape[1] - n_numerical_features
+    else:
+        X = X_numerical
+        n_categorical_features = 0
+    n_variants = len(np.unique(w))
+    if is_classification:
+        y = y_class
+        nuisance_model_factory = LogisticRegression
+    else:
+        y = y_reg
+        nuisance_model_factory = LinearRegression
+
+    ml = RLearner(
+        is_classification,
+        n_variants,
+        nuisance_model_factory=nuisance_model_factory,
+        propensity_model_factory=LGBMClassifier,
+        treatment_model_factory=treatment_model_factory,
+        propensity_model_params={"n_estimators": 1},
+        n_folds=2,
+    )
+    ml.fit(X, y, w)
+
+    onnx_models = []
+    for tv in range(n_variants - 1):
+        model = ml._treatment_models[TREATMENT_MODEL][tv]._overall_estimator
+        onnx_model = onnx_converter(
+            model,
+            initial_types=[
+                (
+                    "X",
+                    FloatTensorType(
+                        [None, n_numerical_features + n_categorical_features]
```

Comment:
It's not super clear to me yet why the end-user builds `onnx_models` themselves. Maybe that'll become clearer with a doc-string for the `build_onnx` method. :)

Reply:
To me there are different reasons for that, the first one being that depending if it's a ``lightgbm``, ``sklearn``, ``xgboost``... a different converter has to be used, obviously we could limit implement this ourselves but then the MetaLearner wouldn't be model-agnostic. Another reason is the fact that this converters have different options so it would require to specify how these options need to be passed. Finally, these converters need the number of features and I think adding a state variable with the number of features won't be correct due to the fact that the feature sets can be different.

One solution would be to pass the converters instead of the models but it would probably require the user to make use of ``partial`` and modify the callables, so I'm not sure what's best.